bliss/catalog.py

```diff
@@ -199,6 +199,68 @@ def gather_param_at_tiles(self, param_name: str, indices: Tensor) -> Tensor:
         idx_to_gather = repeat(indices, "... -> ... k", k=param.size(-1))
         return torch.gather(param, dim=1, index=idx_to_gather)
 
+    def _get_fluxes_of_on_sources(self):
+        """Gets fluxes of "on" sources based on whether the source is a star or galaxy.
+
+        Returns:
+            Tensor: a tensor of fluxes of size (b x nth x ntw x max_sources x 1)
+        """
+        fluxes = torch.where(
+            self["galaxy_bools"], self["galaxy_params"][..., 0, None], self["star_fluxes"]
+        )
+        return torch.where(self.is_on_array[..., None], fluxes, torch.zeros_like(fluxes))
+
+    def get_brightest_source_per_tile(self):
+        """Restrict TileCatalog to only the brightest 'on' source per tile.
+
+        Returns:
+            TileCatalog: a new catalog with only one source per tile
+        """
+        if self.max_sources == 1:
+            return self
+
+        # sort by fluxes of "on" sources to get brightest source per tile
+        on_fluxes = self._get_fluxes_of_on_sources()
+        top_idx = on_fluxes.argsort(dim=3, descending=True)[
+            :, :, :, 0:1, 0
+        ]  # 0:1 keeps dims right for slicing
+
+        d = {}
+        for key, val in self.to_dict().items():
+            if key == "n_sources":
+                d[key] = self.n_sources.bool().int()  # send all positive values to 1, 0 to 0
+            else:
+                param_dim = val.size(-1)
+                idx_to_gather = repeat(top_idx, "... -> ... k", k=param_dim)
+                d[key] = torch.take_along_dim(val, idx_to_gather, dim=3)
+
+        return TileCatalog(self.tile_slen, d)
+
+    def filter_tile_catalog_by_flux(self, min_flux=622, max_flux=1e6):
```

Comment:
Probably better not to set defaults for min/max flux, as the user should be fully aware of what filtering is taking place. Also, there aren't really reasonable defaults that apply across surveys.

Reply:
Would it be better if we make the default range (0, torch.inf)? I was thinking having a default value would be good if you just want to specify either a max or min, like to get all values greater than some threshold.

Reply:
Yes (0, torch.inf) seems good